File: MergeTwoSortedLists.java
// Definition for singly-linked list.
public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
 
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null) return list2;
        else if (list2 == null) return list1;

        ListNode cur1 = list1, cur2 = list2;
        ListNode head = new ListNode();
        ListNode prev = head;
        while (cur1 != null || cur2 != null){
            if (cur1 == null){
                prev.next = cur2;
                prev = cur2;
                cur2 = cur2.next;
            } else if (cur2 == null){
                prev.next = cur1;
                prev = cur1;
                cur1 = cur1.next;
            } else {
                if (cur1.val <= cur2.val){
                    prev.next = cur1;
                    prev = cur1;
                    cur1 = cur1.next;
                } else {
                    prev.next = cur2;
                    prev = cur2;
                    cur2 = cur2.next;
                }
            }
        }
        return head.next;

        // if either of the lists is at the end then use the other list
        // otherwise take the min value and move the pointer of that list up
    }
}

File: MergeTwoSortedLists.cpp
// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (list1 == nullptr) return list2;
        else if (list2 == nullptr) return list1;

        ListNode* head = list1->val <= list2->val ? list1 : list2;
        ListNode* prev = head;
        while (list1 != nullptr || list2 != nullptr){
            ListNode* cur;
            if (list1 == nullptr) {
                cur = list2;
                list2 = list2->next;
            }
            else if (list2 == nullptr) {
                cur = list1;
                list1 = list1->next;
            }
            else {
                if (list1->val <= list2->val){
                    cur = list1;
                    list1 = list1->next;
                } else {
                    cur = list2;
                    list2 = list2->next;
                }
            }
            
            prev->next = cur;
            prev = cur;
        }

        return head;
    }
};

// pretty easy, just like merging two sorted arrays but using pointer logic instead